testutils: add infrastructure for reusable test fixtures

RaduBerinde opened the pull request:
Certain storage-related tests/benchmarks generate fixtures and attempt to reuse them across invocations. This is important because fixtures can be large and slow to generate; but more importantly the generation is non-deterministic and we want to use the exact same fixture when comparing benchmark data.

Currently the tests achieve this by using a `.gitignore`d subdirectory inside the source tree. This does not work with bazel (which executes the test in a sandbox).

This commit addresses the issue by adding new test infrastructure for reusable fixtures. We use the user's `.cache` directory instead of the source tree (specifically $HOME/.cache/crdb-test-fixtures/...). For bazel, we make sure the path is available (and writable) in the sandbox and we pass the path to the test through an env var.

Fixes #83599.

Release note: None
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99624)
<!-- Reviewable:end -->

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

 thanks for doing this! using `make` for these has been a nuisance

> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99624#-:-NRcn7B99XdMGn60cKbc:b-eov1dg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @RaduBerinde, and @rickystewart)

___
*[`pkg/testutils/testfixtures/test_fixtures.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99624#-NRcmMC17kDiasRYY-UK:-NRcmMC17kDiasRYY-UL:b-28854e) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa91cd1a943a8316f702b9063e55b11d062c9fcb/pkg/testutils/testfixtures/test_fixtures.go#L99)):*
> ```Go
> 	tb.Logf("generating fixture %q in %q", name, dir)
> 	generate(dir)
> 	if err := os.WriteFile(completedPath, []byte("foo"), 0755); err != nil {
> ```

maybe should we write the current timestamp to the file?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99624)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @jbowens, and @rickystewart)

___
*[`pkg/testutils/testfixtures/test_fixtures.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99624#-NRcmMC17kDiasRYY-UK:-NRd4rZkDXxWqFUsHum-:b-zhrx6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa91cd1a943a8316f702b9063e55b11d062c9fcb/pkg/testutils/testfixtures/test_fixtures.go#L99)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

maybe should we write the current timestamp to the file?
</blockquote></details>

Hehe, that same thought crossed my mind, but as I was doing it I realized the file will have a modification time anyway. Is there some other info we may want in there?


<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99624)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @RaduBerinde, and @rickystewart)

___
*[`pkg/testutils/testfixtures/test_fixtures.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99624#-NRcmMC17kDiasRYY-UK:-NRdQGst6J5K0sw-FEbm:b-qngwqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa91cd1a943a8316f702b9063e55b11d062c9fcb/pkg/testutils/testfixtures/test_fixtures.go#L99)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hehe, that same thought crossed my mind, but as I was doing it I realized the file will have a modification time anyway. Is there some other info we may want in there?
</blockquote></details>

:) good point. I think ideally we'd have the git SHA that built the fixture, but I don't think we have that available during test builds anymore. Can't think of anything else.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@rickystewart could I get a review of the `cmd/dev` changes?

rickystewart commented on pkg/cmd/dev/test.go:
```diff
@@ -347,6 +347,7 @@ func (d *dev) test(cmd *cobra.Command, commandLine []string) error {
 		args = append(args, "--test_sharding_strategy=disabled")
 	}
 
+	args = append(args, d.getGoTestEnvArgs()...)
```
Looks like this reusable fixture stuff is used for benchmarks rather than unit tests (which is the correct behavior as far as I can tell). Can you please add this extra argument to `dev bench` invocations and NOT `dev test` invocations?

rickystewart reviewed (approved):
Benchmarks run in CI. Do any of the scripts in `build/teamcity` need to updated accordingly to give the benchmarks write-access to the cache directory?

RaduBerinde reviewed (commented):
The bench build inside Bazel Extended CI passed on this PR. I checked the logs from that build and we are using `/home/roach/.cache/`:

```
generating fixture "mvcc-fmtver_10-numKeys_100000-numVersions_2-valueBytes_8-numColFams_0-numRangeKeys_0-transactional_true" in "/home/roach/.cache/crdb-test-fixtures/mvcc-fmtver_10-numKeys_100000-numVersions_2-valueBytes_8-numColFams_0-numRangeKeys_0-transactional_true"
```

But I'm not sure how that works because `bench_impl.sh` doesn't pass the env flag or add .cache to the sandbox.. Any idea how it works? Perhaps they are not running inside a sandbox here?

In any case, note that if we are in a sandbox and the special env var isn't set, we fall back to using a temp dir (which would also be ok behavior for CI).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99624#-:-NRiUo0N83mDyibOqDEZ:b-twjdhr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @jbowens)



<!-- Sent from Reviewable.io -->

rickystewart commented:
> But I'm not sure how that works because bench_impl.sh doesn't pass the env flag or add .cache to the sandbox.. Any idea how it works? Perhaps they are not running inside a sandbox here?

Ah -- so [in CI](https://github.com/cockroachdb/cockroach/blob/master/build/teamcity/cockroach/ci/tests/bench.sh) the benchmarks run with `bazel run`. This is unlike `dev` which uses `bazel test` for the same benchmarks. `bazel test` has this behavior where you cannot write to a directory unless you explicitly allow it, but `bazel run` does not.

RaduBerinde committed:
testutils: add infrastructure for reusable test fixtures

Certain storage-related tests/benchmarks generate fixtures and attempt
to reuse them across invocations. This is important because fixtures
can be large and slow to generate; but more importantly the generation
is non-deterministic and we want to use the exact same fixture when
comparing benchmark data.

Currently the tests achieve this by using a `.gitignore`d subdirectory
inside the source tree. This does not work with bazel (which executes
the test in a sandbox).

This commit addresses the issue by adding new test infrastructure for
reusable fixtures. We use the user's `.cache` directory instead of the
source tree (specifically $HOME/.cache/crdb-test-fixtures/...).
For bazel, we make sure the path is available (and writable) in the
sandbox and we pass the path to the test through an env var.

Fixes #83599.

Release note: None
Epic: none

RaduBerinde reviewed (commented):
Cool, so we should be good. TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99624#-:-NRijwlvBCWv6cBNRWsh:b-hw290e)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @jbowens, and @rickystewart)

___
*[`pkg/cmd/dev/test.go` line 350 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99624#-NRVzDafYYr_1P07TNyP-r1-350:-NRijtFf48M031y7Mdh7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa91cd1a943a8316f702b9063e55b11d062c9fcb/pkg/cmd/dev/test.go#L350)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

Looks like this reusable fixture stuff is used for benchmarks rather than unit tests (which is the correct behavior as far as I can tell). Can you please add this extra argument to `dev bench` invocations and NOT `dev test` invocations? 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
